R/seroprevalence_data.R

```diff
@@ -139,6 +139,32 @@ prepare_bin_data <- function(serodata) {
   return(xx)
 }
 
+#' Function that returns the probability of being seropositive for age-varying
+#' FoI including seroreversion
+#'
+#' @param exposure_age Integer corresponding to the age of the exposed cohort
+#' @param foi Numeric atomic vector corresponding to the age-varying FoI to
+#' simulate from
+#' @param mu Seroreversion rate
+#' @return probability of being seropositive for age-varying FoI
+#' including seroreversion
+get_probability_exact_av <- function(
+    exposure_age,
+    foi,
+    mu = 0
+) {
+  # TODO: Avoid for loop if possible
+  probability <- 0
+  # solves ODE exactly within pieces
+  for (i in 1:exposure_age) {
+    probability <-
+      (1 / (foi[i] + mu)) * exp(- (foi[i] + mu)) *
+      (foi[i] * (exp(foi[i] + mu)  - 1) + probability * (foi[i] + mu)
+      )
```

Comment:
Sorry, I think I may have made a mistake when transcribing my result from Mathematica into R. My result should be the same as eq. (22) in our notes. But I think this isn't true for this R code. I think the code should read:

```r
probability <- (1 / (foi[i] + mu)) * (foi[i] + exp(-(foi[i] + mu)) * (probability * (foi[i] + mu) - foi[i]))
```

Can you check that you agree (I've simplified the expression here and probably should do the same in the notes)?

Reply:
@ben18785 - somehow this new code works the same as what we had previously .. and computes probability the same as the solution we have in the Stan code.

![age_foi_sim_soln](https://github.com/epiverse-trace/serofoi/assets/7207204/043a5046-4d83-4007-ba43-ed04243dd9db)

Reply:
It seems like they're all the same. I rewrote it as:

```math
X(t) = X_0 e^{-(\lambda + \mu)} + \frac{\lambda}{\lambda + \mu} (1 - e^{-(\lambda + \mu)}),
```
which is equivalent to what you originally coded:

```math
X(t) = \frac{e^{-(\lambda + \mu)}}{\lambda + \mu} \left( \lambda (e^{\lambda + \mu} - 1) + X_0 (\lambda + \mu) \right).
```
You must have factorized the exponential and the denominator to simplify. Introducing the exponential in the parenthesis and factorizing it inside (leaving out $\lambda$) yields your current suggestion:

```math
X(t) = \frac{1}{\lambda + \mu} \left(  \lambda + e^{-(\lambda + \mu)} (X_0 (\lambda + \mu) - \lambda) \right)
```

I computed them with $\mu = 0.1$ and they all coincide anyway:
![image](https://github.com/epiverse-trace/serofoi/assets/45337127/7eaf35cf-3db1-42a5-9ce2-bcc498da7e16)

I think the last one is better just because it uses only one exponential,  but please let me know which version seems less convoluted to you.

Btw, I checked that your solution actually solves the ODE by direct derivation just in case. I haven't gone through the whole demonstration assuming yearly constant FOI just yet, but I understand better your argument now.

Reply:
Great, thanks @ekamau and @ntorresd and apologies for my convoluted thinking here. I agree that the one with only the single exponential is best.